### PR TITLE
Issue/277

### PR DESCRIPTION
**Pull request by minborg**
This PR introduces additional tests and checkings around negative values. This prevents a number of bugs and JVM crashes in the tests.

### CODE REVIEW

**Review by RobAustin — APPROVED** [review]

File: src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java [hunk 1]
```diff
@@ -273,12 +277,14 @@ public long capacity() {
     @Override
     public long addressForRead(long offset)
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java:278]: Should we add @NonNegative annotation to these for consistency?

File: src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java [hunk 2]
```diff
@@ -273,12 +277,14 @@ public long capacity() {
...
             throws UnsupportedOperationException, IllegalStateException, BufferUnderflowException {
+        requireNonNegative(offset);
         return base.addressForRead(offset);
     }
 
     @Override
     public long addressForWrite(long offset)
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java:285]: And here?

File: src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java [hunk 3]
```diff
@@ -881,6 +896,7 @@ public void readWithLength(Bytes bytes)
     @NotNull
     public Bytes<Void> writePosition(long position)
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java:897]: And here?

File: src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java [hunk 4]
```diff
@@ -1106,11 +1122,11 @@ private void doIndent() {
 
     @Override
     @NotNull
-    public Bytes<Void> write(@NotNull byte[] bytes, int offset, int length)
+    public Bytes<Void> write(@NotNull byte[] byteArray, int offset, int length)
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java:1125]: Why do we use @NotNull for byte arrays sometimes and not other times (even though we assert not null)

File: src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java [hunk 5]
```diff
@@ -1655,6 +1674,9 @@ public Bytes write(@NotNull InputStream inputStream)
     @Override
     public Bytes<Void> write(@NotNull BytesStore bytes, long offset, long length)
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java:1675]: @nonnegative

File: src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java [hunk 6]
```diff
@@ -1709,6 +1731,8 @@ public Bytes write(@NotNull InputStream inputStream)
     @Override
     public void writePositionRemaining(long position, long length)
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java:1732]: @nonnegative

**Review by nicktindall — APPROVED** [review]

**Comment by nicktindall** [comment]
Seemed to be some inconsistencies with our use of the annotations, but other than that LGTM!

File: src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java (hunk 1 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java:278]: Agree. I believe that offsets should always be from the start of the Bytes and thus always be >= 0

File: src/main/java/net/openhft/chronicle/bytes/internal/NativeBytesStore.java [hunk 7]
```diff
@@ -515,11 +519,15 @@ public long readVolatileLong(long offset) {
...
-    public NativeBytesStore<U> write(
-            long offsetInRDO, byte[] bytes, int offset, int length)
-            throws IllegalStateException {
-        requireNonNull(bytes);
-        memory.copyMemory(bytes, offset, address + translate(offsetInRDO), length);
+    public NativeBytesStore<U> write(@NonNegative final long offsetInRDO,
+                                     final byte[] byteArray,
+                                     @NonNegative final int offset,
+                                     @NonNegative final  int length) throws IllegalStateException {
+        requireNonNegative(offsetInRDO);
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/bytes/internal/NativeBytesStore.java:526]: My dream is that we (and customers) can validate `@NonNegative` etc at compile time using static analysis, and we can turn off checks at runtime

**Review by JerryShea — APPROVED** [review]
I like the additional tests and rigour but I would be much happier if we bolstered our performance tests for common use cases

File: src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java (hunk 1 above)
Comment (minborg) [reply src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java:278]: I think we should. This makes the contract clearer. I have raised https://github.com/OpenHFT/Chronicle-Bytes/issues/290

File: src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java (hunk 4 above)
Comment (minborg) [reply src/main/java/net/openhft/chronicle/bytes/HexDumpBytes.java:1125]: Originally there was a misunderstanding about what `@NotNull` means for arrays. I raised a new issue about this:
https://github.com/OpenHFT/Chronicle-Bytes/issues/291

File: src/main/java/net/openhft/chronicle/bytes/internal/NativeBytesStore.java (hunk 7 above)
Comment (minborg) [reply src/main/java/net/openhft/chronicle/bytes/internal/NativeBytesStore.java:526]: This is absolutely doable. We could clone and modify the current non-null plugin.

**Pull request merged**